### PR TITLE
Create command to export images from .DAT file

### PR DESCRIPTION
**Pull request by Gymnasiast**

### CODE REVIEW

File: src/openrct2/CmdlineSprite.cpp [hunk 1]
```diff
@@ -371,7 +374,8 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
             return -1;
         }
 
-        if (!sprite_file_export(spriteIndex, outputPath))
+        rct_g1_element* spriteHeader = &spriteFileEntries[spriteIndex];
```
Comment (tupaschoal) [inline src/openrct2/CmdlineSprite.cpp:377]:
```suggestion
        spriteHeader = &spriteFileEntries[spriteIndex];
```
`spriteHeader` already exists (it is a parameter) so having the variable here again is probably shadowing it.

File: src/openrct2/CmdlineSprite.cpp [hunk 2]
```diff
@@ -431,7 +435,8 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
                 printf("\r%d / %d, %d%%", spriteIndex, maxIndex, spriteIndex / maxIndex);
             }
 
-            if (!sprite_file_export(spriteIndex, outputPath))
+            rct_g1_element* spriteHeader = &spriteFileEntries[spriteIndex];
```
Comment (tupaschoal) [inline src/openrct2/CmdlineSprite.cpp:438]: Same

File: src/openrct2/CmdlineSprite.cpp [hunk 3]
```diff
@@ -452,6 +457,157 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
...
+            return -1;
+        }
+
+        char datName[9] = { 0 };
+        std::fill_n(datName, 8, ' ');
+        int32_t i = 0;
+        for (const char* ch = argv[1]; *ch != '\0' && i < 8; ch++)
+        {
+            datName[i++] = *ch;
+        }
```
Comment (tupaschoal) [inline src/openrct2/CmdlineSprite.cpp:474]: Would doing `safe_strcpy(outputPath, argv[2], MAX_PATH);` like the other place work?

File: src/openrct2/CmdlineSprite.cpp [hunk 4]
```diff
@@ -452,6 +457,157 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
...
+                break;
+            }
+            case OBJECT_TYPE_SMALL_SCENERY:
+            case OBJECT_TYPE_LARGE_SCENERY:
+            case OBJECT_TYPE_WALLS:
+            case OBJECT_TYPE_BANNERS:
+            case OBJECT_TYPE_PATH_BITS:
+            {
+                auto& objMgr = OpenRCT2::GetContext()->GetObjectManager();
+                auto obj = objMgr.GetLoadedObject(objectType, entryIndex);
```
Comment (tupaschoal) [inline src/openrct2/CmdlineSprite.cpp:526]: You already have `objManager` above, just reuse it
```suggestion
                auto obj = objManager.GetLoadedObject(objectType, entryIndex);
```

File: src/openrct2/CmdlineSprite.cpp [hunk 5]
```diff
@@ -452,6 +457,157 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
...
+            case OBJECT_TYPE_SCENERY_GROUP:
+            {
+                auto sceneryGroupEntry = get_scenery_group_entry(entryIndex);
+                imagesOffset = sceneryGroupEntry->image;
+                break;
+            }
+            case OBJECT_TYPE_PARK_ENTRANCE:
+            {
+                auto& objMgr = OpenRCT2::GetContext()->GetObjectManager();
+                auto obj = objMgr.GetLoadedObject(objectType, entryIndex);
```
Comment (tupaschoal) [inline src/openrct2/CmdlineSprite.cpp:549]:
```suggestion
                auto obj = objManager.GetLoadedObject(objectType, entryIndex);
```

File: src/openrct2/CmdlineSprite.cpp [hunk 6]
```diff
@@ -452,6 +457,157 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
...
+                break;
+            }
+            default:
+            {
+                fprintf(stderr, "Cannot extract images from this type of object.\n");
+                return -1;
+            }
+        }
+
+        int32_t numbers = std::max(1, static_cast<int32_t>(std::floor(std::log(maxIndex))));
```
Comment (tupaschoal) [inline src/openrct2/CmdlineSprite.cpp:564]: `numbers` seems a bit generic

File: src/openrct2/CmdlineSprite.cpp (hunk 1 above)
Comment (Gymnasiast) [reply src/openrct2/CmdlineSprite.cpp:377]: I don't know what you mean. `spriteHeader` is not a parameter of `cmdline_for_sprite()`.

File: src/openrct2/CmdlineSprite.cpp (hunk 6 above)
Comment (Gymnasiast) [reply src/openrct2/CmdlineSprite.cpp:564]: It was copied from the other function. It saves how many digits there are in the total number of images. Perhaps `numDigits` is better?

File: src/openrct2/CmdlineSprite.cpp (hunk 3 above)
Comment (Gymnasiast) [reply src/openrct2/CmdlineSprite.cpp:474]: No. This concerns a .DAT name, consisting of 8 characters, not a full path.

File: src/openrct2/CmdlineSprite.cpp (hunk 1 above)
Comment (tupaschoal) [reply src/openrct2/CmdlineSprite.cpp:378]: The folding fooled me and I thought this was still in `sprite_file_export`

File: src/openrct2/CmdlineSprite.cpp (hunk 6 above)
Comment (tupaschoal) [reply src/openrct2/CmdlineSprite.cpp]: Does look better, IMHO

File: src/openrct2/CmdlineSprite.cpp [hunk 7]
```diff
@@ -450,6 +458,155 @@ int32_t cmdline_for_sprite(const char** argv, int32_t argc)
         sprite_file_close();
         return 1;
     }
+    else if (_strcmpi(argv[0], "exportalldat") == 0)
+    {
+        if (argc < 3)
+        {
+            fprintf(stdout, "usage: sprite exportall <spritefile> <output directory>\n");
```
Comment (AaronVanGeffen) [inline src/openrct2/CmdlineSprite.cpp:465]: Should this have read `exportalldat`?

Comment (Gymnasiast) [reply src/openrct2/CmdlineSprite.cpp:465]: It should have, yes.

https://github.com/OpenRCT2/OpenRCT2/pull/11799

**Pull request merged**